lib/fs.js

```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
   this.closed = false;
 
   if (this.start !== undefined) {
-    if (typeof this.start !== 'number' || Number.isNaN(this.start)) {
+    if (typeof this.start !== 'number') {
       throw new ERR_INVALID_ARG_TYPE('start', 'number', this.start);
     }
-    if (this.end === undefined) {
-      this.end = Infinity;
-    } else if (typeof this.end !== 'number' || Number.isNaN(this.end)) {
-      throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+
+    if (Number.isNaN(this.start)) {
+      throw new ERR_OUT_OF_RANGE('start', 'not NaN', this.start);
     }
 
-    if (this.start > this.end) {
-      const errVal = `{start: ${this.start}, end: ${this.end}}`;
-      throw new ERR_OUT_OF_RANGE('start', '<= "end"', errVal);
+    if (this.start < 0) {
+      throw new ERR_OUT_OF_RANGE('start', 'positive', this.start);
+    }
+
+    if (!Number.isSafeInteger(this.start)) {
+      throw new ERR_OUT_OF_RANGE('start', 'integer', this.start);
     }
 
     this.pos = this.start;
   }
 
-  // Backwards compatibility: Make sure `end` is a number regardless of `start`.
-  // TODO(addaleax): Make the above typecheck not depend on `start` instead.
-  // (That is a semver-major change).
-  if (typeof this.end !== 'number')
+  if (this.end === undefined) {
     this.end = Infinity;
-  else if (Number.isNaN(this.end))
-    throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+  } else if (this.end !== Infinity) {
+    if (typeof this.end !== 'number') {
+      throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+    }
+
+    if (Number.isNaN(this.end)) {
+      throw new ERR_OUT_OF_RANGE('end', 'not NaN', this.end);
+    }
+
+    if (this.end < 0) {
+      throw new ERR_OUT_OF_RANGE('end', 'positive', this.end);
+    }
+
+    if (!Number.isSafeInteger(this.end)) {
+      throw new ERR_OUT_OF_RANGE('end', 'integer', this.end);
```

Comment:
This is now not specific enough anymore. E.g., `2 ** 54` would now report it is not an integer. The same above.